Feat: Better complex struct inputs

technophile-04 opened the pull request:
## Description
https://github.com/scaffold-eth/scaffold-eth-2/assets/80153681/c388aa58-ce10-4089-9d49-e633ba83524f

Tried handling better structs inputs and complex structs array, lol not the best optimized approach but in my testing seems to work nicely 

For testing I have updated `YourContract.sol` with a nested struct along with a multidimensional struct array but please feel free to play around / test different cases. 

You could also try putting [Opensea contract](https://etherscan.io/address/0x00000000006c3852cbEf3e08E8dF289169EdE581#code) in externalContract on mainnet to see how UI looks for testing. 

Didn't go too much on visual UI part but maybe we could handle it more better just wanted to get functionality inplace.

Also note, this PR only handles structs array and not primitive types array maybe we could it another PR since it should be alot easier and didn't wanted to clutter this PR 🙌

OG Idea: https://twitter.com/_jxom/status/1748068858047668227

rin-st commented on packages/nextjs/app/debug/_components/contract/utilsContract.tsx:
```diff
@@ -49,4 +71,79 @@ const getInitialFormState = (abiFunction: AbiFunction) => {
+  const initialForm: Record<string, any> = {};
+  if (abiTupleParameter.components.length === 0) return initialForm;
+  abiTupleParameter.components.forEach((component, componentIndex) => {
+    const key = getFunctionInputKey("0_" + abiTupleParameter.name || "tuple", component, componentIndex);
```
why "0_"?

technophile-04 replied:
Actually, this function constructs the initial state for `TupleArray` and since initially while rendering `TupleArray`we are displaying 0th element hence a state for it. 

<details>
<summary>
More detailed explanation 
</summary>

Lets suppose : 

```solidity
struct SimpleStruct {
  uint x;
  uint y;
}

// ---

struct NestedStruct {
  uint a;
  SimpleStruct[] b;
}

```

Then the inital form state  for `b`  `TupleArray` will look like : 

```js
{
    "0_b_x_uint256_uint256": "",
    "0_b_y_uint256_uint256": ""
}
```
since default are showing 0th element

</details>

rin-st commented:
Magic! 🪄 
Looks awesome! Really great job!

Noticed that I can't send array with 0 entries, there's at least one and I can't remove it.
Also a question: do we need it also for reading, or only for writing?

technophile-04 commented:
> Noticed that I can't send array with 0 entries, there's at least one and I can't remove it.

Ohh yeah, I completely forgot that you could pass in empty array fixed at https://github.com/scaffold-eth/scaffold-eth-2/pull/702/commits/df3fd3eb1b7c46bf24697d27538a996a4ff6b5d8, by default it shows 0th element (I think its nice UX to show it at default, but we could keep empty by default) but you can remove 0th element and empty array will passed

 > Also a question: do we need it also for reading, or only for writing?
 
 Yeah lol again forgot about read ! fixed at https://github.com/scaffold-eth/scaffold-eth-2/pull/702/commits/1394397f3f1f738573d6053ec25e029c5337aecc also updated test contract 

Thanks Rinat 🙌

rin-st reviewed (approved):
Awesome, thanks!

Pabl0cks reviewed (approved):
This is just GREAT!! Awesome job implementing it @technophile-04 🔥🔥

I like UI a lot! The only thing I felt that could be a bit confusing to some users was the visual representation for each tuple item (0, 1, 2..). Maybe we could add a bit more descriptive text like Item[0], or TupleItem[0] or something similar. 

<details>
<summary>If we want to add more visual differentiation between tuple items we could play with bg color. Here is an example with zebra pattern, but maybe that's a bit ugly, this could go in a future iteration with some designer help :)
</summary>


![image](https://github.com/scaffold-eth/scaffold-eth-2/assets/55535804/b1713c2f-744d-4157-a072-a2c20b1c0c78)

</details>

technophile-04 commented:
> visual representation for each tuple item (0, 1, 2..). Maybe we could add a bit more descriptive text like Item[0], or TupleItem[0] or something similar.

Yup, completely get it,  updated at https://github.com/scaffold-eth/scaffold-eth-2/pull/702/commits/cc89cfead4c5175ba52ff83ea895a6da30319c32

![Screenshot 2024-02-13 at 11 28 13 PM](https://github.com/scaffold-eth/scaffold-eth-2/assets/80153681/a74f7171-e235-4d3d-a143-d10f83623a3a)

- Used `tuple` instead `item` or any other name since people can name the args as item or any other thing and it will clash with our naming and might make it more confusing (they won't be able to use `tuple` as arg name since its reserved)
- Only showing `tuple[index]` for 1D array elements since technically marking 2D / ND  marking their all levels with tuple[index] might not make sense 


> to add more visual differentiation between tuple items we could play with bg color. Here is an example with a zebra pattern, but maybe that's a bit ugly, this could go in a future iteration with some designer help

Love this idea !!!! It kind of makes it visually easy to digest, we can think maybe changing the color a bit for an alternate zebra strip, not adding this changes in PR but should probably handle this in separate PR  🙌

Tysm Pablo for the great feedback !!

technophile-04 commented:
Tysm all for testing !! Just locked dasiyUI version at https://github.com/scaffold-eth/scaffold-eth-2/pull/702/commits/f9c4a0f6bb849d58348bb78a7765049f01a4a1d5 since they seem to have broken Accordian in their latest version and when you deploy to prod we use their version and its broken there eg: 

![Screenshot 2024-02-14 at 6 56 45 PM](https://github.com/scaffold-eth/scaffold-eth-2/assets/80153681/8028d155-7057-4dc7-879c-9a890e4f33ad)

The accordian arrow does not show up here https://se2-structs.vercel.app/

After locking daisyUI version here is prod version : https://se2strcut-prod.vercel.app/

TODO: 
1. Create an issue on SE-2 for better styling of this https://github.com/scaffold-eth/scaffold-eth-2/pull/702#pullrequestreview-1875065425
2. Create an issue on DaisyUI regarding Accordian issue 

Merging this 🙌
